File: backend/app/services/generation_service.py
import uuid
import logging
from datetime import datetime, timezone
from typing import List, Dict, Any, Optional

from app.schemas.generation import (
    GenerationRequest,
    GeneratedDeliverable,
    GenerationResponse,
    ALLOWED_DELIVERABLES,
)
from app.services.prompt_builder import (
    build_system_instruction,
    build_transformation_prompt,
    DELIVERABLE_SPECIFICATIONS,
)
from app.services.gemini_service import get_gemini_service, GeminiService
from app.core.config import settings

logger = logging.getLogger(__name__)


def _format_markdown_fallback(deliverable_id: str, data: Dict[str, Any], title: str) -> str:
    """Helper to synthesize a readable markdown representation if model returns structured fields."""
    lines = [f"# {title}\n"]

    if deliverable_id == "executive_summary":
        if "summary" in data:
            lines.append(f"## Executive Overview\n{data['summary']}\n")
        if "key_points" in data and isinstance(data["key_points"], list):
            lines.append("## Key Strategic Points")
            for p in data["key_points"]:
                lines.append(f"- {p}")
            lines.append("")
        if "important_findings" in data and isinstance(data["important_findings"], list):
            lines.append("## Important Findings")
            for f in data["important_findings"]:
                lines.append(f"- {f}")
            lines.append("")
        if "recommended_actions" in data and isinstance(data["recommended_actions"], list):
            lines.append("## Recommended Actions")
            for a in data["recommended_actions"]:
                lines.append(f"- {a}")

    elif deliverable_id == "linkedin_post":
        if "hook" in data:
            lines.append(f"{data['hook']}\n")
        if "body" in data:
            lines.append(f"{data['body']}\n")
        if "call_to_action" in data:
            lines.append(f"👉 {data['call_to_action']}\n")
        if "hashtags" in data and isinstance(data["hashtags"], list):
            lines.append(" ".join(data["hashtags"]))

    elif deliverable_id == "twitter_post":
        if "thread_posts" in data and isinstance(data["thread_posts"], list):
            for i, tweet in enumerate(data["thread_posts"], 1):
                lines.append(f"**Post {i}/{len(data['thread_posts'])}**\n{tweet}\n")
        if "key_takeaway" in data:
            lines.append(f"💡 Key Takeaway: {data['key_takeaway']}\n")
        if "hashtags" in data and isinstance(data["hashtags"], list):
            lines.append(" ".join(data["hashtags"]))

    elif deliverable_id == "advisory":
        if "context" in data:
            lines.append(f"## Context & Background\n{data['context']}\n")
        if "key_information" in data and isinstance(data["key_information"], list):
            lines.append("## Key Information")
            for info in data["key_information"]:
                lines.append(f"- {info}")
            lines.append("")
        if "action_items" in data and isinstance(data["action_items"], list):
            lines.append("## Action Items")
            for act in data["action_items"]:
                lines.append(f"- {act}")
            lines.append("")
        if "cautions_or_notes" in data and isinstance(data["cautions_or_notes"], list):
            lines.append("## Cautions & Notes")
            for c in data["cautions_or_notes"]:
                lines.append(f"- ⚠️ {c}")

    elif deliverable_id == "infographic":
        if "core_message" in data:
            lines.append(f"**Core Message**: {data['core_message']}\n")
        if "key_facts_and_metrics" in data and isinstance(data["key_facts_and_metrics"], list):
            lines.append("## Anchor Metrics & Facts")
            for metric in data["key_facts_and_metrics"]:
                lines.append(f"- 📊 {metric}")
            lines.append("")
        if "sections" in data and isinstance(data["sections"], list):
            lines.append("## Infographic Sections")
            for s in data["sections"]:
                heading = s.get("heading", "Section")
                content = s.get("content", "")
                visual = s.get("visual_cue", "")
                lines.append(f"### {heading}")
                lines.append(f"{content}")
                if visual:
                    lines.append(f"*Visual Direction: {visual}*\n")
        if "visual_layout_guidance" in data:
            lines.append(f"## Visual Layout Guidance\n{data['visual_layout_guidance']}")

    elif deliverable_id == "presentation":
        if "slides" in data and isinstance(data["slides"], list):
            for slide in data["slides"]:
                num = slide.get("slide_number", "")
                stitle = slide.get("slide_title", "Slide")
                lines.append(f"## Slide {num}: {stitle}")
                bullets = slide.get("bullet_points", [])
                if isinstance(bullets, list):
                    for b in bullets:
                        lines.append(f"- {b}")
                if "visual_concept" in slide:
                    lines.append(f"\n*Visual Concept: {slide['visual_concept']}*")
                if "speaker_notes" in slide:
                    lines.append(f"\n> **Speaker Notes:** {slide['speaker_notes']}\n")

    elif deliverable_id == "video_package":
        if "duration_guidance" in data:
            lines.append(f"**Target Duration:** {data['duration_guidance']}\n")
        if "scenes" in data and isinstance(data["scenes"], list):
            lines.append("## Scene Breakdown")
            for scene in data["scenes"]:
                snum = scene.get("scene_number", "")
                lines.append(f"### Scene {snum}")
                if "visual_direction" in scene:
                    lines.append(f"**Visual:** {scene['visual_direction']}")
                if "narration_script" in scene:
                    lines.append(f"**Narration:** \"{scene['narration_script']}\"")
                if "on_screen_text" in scene:
                    lines.append(f"**On-Screen Text:** {scene['on_screen_text']}\n")
        if "subtitles" in data:
            lines.append(f"## Full Subtitles Transcript\n{data['subtitles']}")

    else:
        # Generic fallback string formatting
        for k, v in data.items():
            lines.append(f"**{k.replace('_', ' ').title()}**: {v}")

    return "\n".join(lines).strip()


class GenerationService:
    """Orchestrates content transformation requests with Gemini AI."""

    def __init__(self, gemini_service: Optional[GeminiService] = None):
        self._gemini_service = gemini_service

    def _get_gemini(self) -> GeminiService:
        if self._gemini_service is not None:
            return self._gemini_service
        return get_gemini_service()

    def execute_transformation(self, request: GenerationRequest) -> GenerationResponse:
        """Execute transformation and synthesize structured deliverables."""
        session_id = f"gen_session_{uuid.uuid4().hex[:12]}"
        now_iso = datetime.now(timezone.utc).isoformat()

        # Build prompt & system instructions
        system_instruction = build_system_instruction()
        prompt = build_transformation_prompt(request)

        gemini = self._get_gemini()

        try:
            raw_result = gemini.generate_json_content(
                prompt=prompt,
                system_instruction=system_instruction,
            )
        except Exception as exc:
            logger.error(f"Generation execution failed: {str(exc)}")
            return GenerationResponse(
                success=False,
                sessionId=session_id,
                status="failed",
                model=settings.GEMINI_MODEL,
                deliverables=[],
                error=str(exc),
                generatedAt=now_iso,
            )

        # Parse and validate the returned deliverables against requested IDs
        parsed_deliverables: List[GeneratedDeliverable] = []
        raw_deliverables_list = raw_result.get("deliverables", []) if isinstance(raw_result, dict) else []

        # Create mapping of deliverableId -> item
        deliverables_by_id: Dict[str, Dict[str, Any]] = {}
        if isinstance(raw_deliverables_list, list):
            for item in raw_deliverables_list:
                if isinstance(item, dict) and "deliverableId" in item:
                    deliverables_by_id[str(item["deliverableId"])] = item

        for d_id in request.deliverables:
            spec = DELIVERABLE_SPECIFICATIONS.get(d_id, {"name": d_id.replace("_", " ").title()})
            default_title = spec.get("name", d_id.title())

            if d_id in deliverables_by_id:
                raw_item = deliverables_by_id[d_id]
                title = raw_item.get("title") or default_title
                structured_data = raw_item.get("structuredData")
                if not isinstance(structured_data, dict):
                    structured_data = {}

                # If content is provided and is string, use it; otherwise generate markdown from structuredData
                content = raw_item.get("content")
                if not content or not isinstance(content, str) or not content.strip():
                    content = _format_markdown_fallback(d_id, structured_data, title)
                else:
                    content = content.strip()

                parsed_deliverables.append(
                    GeneratedDeliverable(
                        deliverableId=d_id,
                        title=title,
                        content=content,
                        structuredData=structured_data,
                        status="completed",
                    )
                )
            else:
                # Deliverable was requested but missing in model output
                logger.warning(f"Deliverable '{d_id}' requested but missing in Gemini output payload.")
                parsed_deliverables.append(
                    GeneratedDeliverable(
                        deliverableId=d_id,
                        title=default_title,
                        content=f"Deliverable generation could not be completed for {default_title}.",
                        structuredData=None,
                        status="failed",
                        error=f"Model output was missing requested deliverable '{d_id}'.",
                    )
                )

        completed_count = sum(1 for d in parsed_deliverables if d.status == "completed")
        overall_status = "completed" if completed_count == len(request.deliverables) else ("partial" if completed_count > 0 else "failed")

        return GenerationResponse(
            success=completed_count > 0,
            sessionId=session_id,
            status=overall_status,
            model=settings.GEMINI_MODEL,
            deliverables=parsed_deliverables,
            error=None if overall_status == "completed" else f"Completed {completed_count}/{len(request.deliverables)} deliverables.",
            generatedAt=now_iso,
        )


# Singleton instance
_generation_service_instance: Optional[GenerationService] = None


def get_generation_service() -> GenerationService:
    global _generation_service_instance
    if _generation_service_instance is None:
        _generation_service_instance = GenerationService()
    return _generation_service_instance

File: server/generationService.ts
import {
  GenerationRequestPayload,
  DELIVERABLE_SPECIFICATIONS,
  buildSystemInstruction,
  buildTransformationPrompt,
} from "./promptBuilder";
import { generateStructuredJson } from "./geminiService";

export interface GeneratedDeliverableResponse {
  deliverableId: string;
  title: string;
  content: string;
  structuredData?: Record<string, any> | null;
  status: string;
  error?: string | null;
}

export interface GenerationResponsePayload {
  success: boolean;
  sessionId: string;
  status: "completed" | "partial" | "failed";
  model?: string;
  deliverables: GeneratedDeliverableResponse[];
  error?: string | null;
  generatedAt: string;
}

function formatMarkdownFallback(deliverableId: string, data: Record<string, any>, title: string): string {
  const lines: string[] = [`# ${title}\n`];

  if (deliverableId === "executive_summary") {
    if (data.summary) lines.push(`## Executive Overview\n${data.summary}\n`);
    if (Array.isArray(data.key_points)) {
      lines.push("## Key Strategic Points");
      data.key_points.forEach((p: string) => lines.push(`- ${p}`));
      lines.push("");
    }
    if (Array.isArray(data.important_findings)) {
      lines.push("## Important Findings");
      data.important_findings.forEach((f: string) => lines.push(`- ${f}`));
      lines.push("");
    }
    if (Array.isArray(data.recommended_actions)) {
      lines.push("## Recommended Actions");
      data.recommended_actions.forEach((a: string) => lines.push(`- ${a}`));
    }
  } else if (deliverableId === "linkedin_post") {
    if (data.hook) lines.push(`${data.hook}\n`);
    if (data.body) lines.push(`${data.body}\n`);
    if (data.call_to_action) lines.push(`👉 ${data.call_to_action}\n`);
    if (Array.isArray(data.hashtags)) lines.push(data.hashtags.join(" "));
  } else if (deliverableId === "twitter_post") {
    if (Array.isArray(data.thread_posts)) {
      data.thread_posts.forEach((t: string, idx: number) => {
        lines.push(`**Post ${idx + 1}/${data.thread_posts.length}**\n${t}\n`);
      });
    }
    if (data.key_takeaway) lines.push(`💡 Key Takeaway: ${data.key_takeaway}\n`);
    if (Array.isArray(data.hashtags)) lines.push(data.hashtags.join(" "));
  } else if (deliverableId === "advisory") {
    if (data.context) lines.push(`## Context & Background\n${data.context}\n`);
    if (Array.isArray(data.key_information)) {
      lines.push("## Key Information");
      data.key_information.forEach((k: string) => lines.push(`- ${k}`));
      lines.push("");
    }
    if (Array.isArray(data.action_items)) {
      lines.push("## Action Items");
      data.action_items.forEach((a: string) => lines.push(`- ${a}`));
      lines.push("");
    }
    if (Array.isArray(data.cautions_or_notes)) {
      lines.push("## Cautions & Notes");
      data.cautions_or_notes.forEach((c: string) => lines.push(`- ⚠️ ${c}`));
    }
  } else if (deliverableId === "infographic") {
    if (data.core_message) lines.push(`**Core Message**: ${data.core_message}\n`);
    if (Array.isArray(data.key_facts_and_metrics)) {
      lines.push("## Anchor Metrics & Facts");
      data.key_facts_and_metrics.forEach((m: string) => lines.push(`- 📊 ${m}`));
      lines.push("");
    }
    if (Array.isArray(data.sections)) {
      lines.push("## Infographic Sections");
      data.sections.forEach((s: any) => {
        lines.push(`### ${s.heading || "Section"}`);
        if (s.content) lines.push(`${s.content}`);
        if (s.visual_cue) lines.push(`*Visual Direction: ${s.visual_cue}*\n`);
      });
    }
    if (data.visual_layout_guidance) {
      lines.push(`## Visual Layout Guidance\n${data.visual_layout_guidance}`);
    }
  } else if (deliverableId === "presentation") {
    if (Array.isArray(data.slides)) {
      data.slides.forEach((s: any) => {
        lines.push(`## Slide ${s.slide_number || ""}: ${s.slide_title || "Slide"}`);
        if (Array.isArray(s.bullet_points)) {
          s.bullet_points.forEach((b: string) => lines.push(`- ${b}`));
        }
        if (s.visual_concept) lines.push(`\n*Visual Concept: ${s.visual_concept}*`);
        if (s.speaker_notes) lines.push(`\n> **Speaker Notes:** ${s.speaker_notes}\n`);
      });
    }
  } else if (deliverableId === "video_package") {
    if (data.duration_guidance) lines.push(`**Target Duration:** ${data.duration_guidance}\n`);
    if (Array.isArray(data.scenes)) {
      lines.push("## Scene Breakdown");
      data.scenes.forEach((sc: any) => {
        lines.push(`### Scene ${sc.scene_number || ""}`);
        if (sc.visual_direction) lines.push(`**Visual:** ${sc.visual_direction}`);
        if (sc.narration_script) lines.push(`**Narration:** "${sc.narration_script}"`);
        if (sc.on_screen_text) lines.push(`**On-Screen Text:** ${sc.on_screen_text}\n`);
      });
    }
    if (data.subtitles) lines.push(`## Full Subtitles Transcript\n${data.subtitles}`);
  } else {
    for (const [k, v] of Object.entries(data)) {
      lines.push(`**${k.replace(/_/g, " ")}**: ${typeof v === "object" ? JSON.stringify(v) : v}`);
    }
  }

  return lines.join("\n").trim();
}

export async function executeTransformation(request: GenerationRequestPayload): Promise<GenerationResponsePayload> {
  const sessionId = `gen_session_${Math.random().toString(36).substring(2, 14)}`;
  const nowIso = new Date().toISOString();

  if (!request.sourceText || !request.sourceText.trim()) {
    throw new Error("Source text cannot be empty.");
  }
  if (!request.deliverables || request.deliverables.length === 0) {
    throw new Error("At least one target deliverable must be selected.");
  }

  const systemInstruction = buildSystemInstruction();
  const prompt = buildTransformationPrompt(request);

  let rawResult: any;
  let modelUsed = "gemini-3.6-flash";

  try {
    const genResult = await generateStructuredJson(prompt, systemInstruction);
    rawResult = genResult.data;
    modelUsed = genResult.modelUsed;
  } catch (err: any) {
    console.error("[GenerationService] Execution failed:", err);
    return {
      success: false,
      sessionId,
      status: "failed",
      model: modelUsed,
      deliverables: [],
      error: err.message || "Failed to generate deliverables with Gemini AI.",
      generatedAt: nowIso,
    };
  }

  const parsedDeliverables: GeneratedDeliverableResponse[] = [];
  const rawList = Array.isArray(rawResult?.deliverables) ? rawResult.deliverables : [];
  const deliverablesById: Record<string, any> = {};

  for (const item of rawList) {
    if (item && item.deliverableId) {
      deliverablesById[String(item.deliverableId)] = item;
    }
  }

  for (const dId of request.deliverables) {
    const spec = DELIVERABLE_SPECIFICATIONS[dId] || { name: dId.replace(/_/g, " ") };
    const defaultTitle = spec.name;

    if (deliverablesById[dId]) {
      const rawItem = deliverablesById[dId];
      const title = rawItem.title || defaultTitle;
      const structuredData = typeof rawItem.structuredData === "object" ? rawItem.structuredData : {};
      let content = typeof rawItem.content === "string" && rawItem.content.trim() ? rawItem.content.trim() : "";
      if (!content) {
        content = formatMarkdownFallback(dId, structuredData, title);
      }

      parsedDeliverables.push({
        deliverableId: dId,
        title,
        content,
        structuredData,
        status: "completed",
      });
    } else {
      parsedDeliverables.push({
        deliverableId: dId,
        title: defaultTitle,
        content: `Deliverable generation could not be completed for ${defaultTitle}.`,
        structuredData: null,
        status: "failed",
        error: `Model output did not include deliverable '${dId}'.`,
      });
    }
  }

  const completedCount = parsedDeliverables.filter((d) => d.status === "completed").length;
  const overallStatus =
    completedCount === request.deliverables.length
      ? "completed"
      : completedCount > 0
      ? "partial"
      : "failed";

  return {
    success: completedCount > 0,
    sessionId,
    status: overallStatus,
    model: modelUsed,
    deliverables: parsedDeliverables,
    error: overallStatus === "completed" ? null : `Completed ${completedCount}/${request.deliverables.length} deliverables.`,
    generatedAt: nowIso,
  };
}
